demos/benchmark/embeddings/benchmark_embeddings.py

```diff
@@ -0,0 +1,178 @@
+
+print("Number of documents:",len(docs))
+
+batch_size = args['batch_size']
+
+def get_tokens(docs, model):
+    tokenizer = AutoTokenizer.from_pretrained(model)
+    documents = docs.iter(batch_size=1)
+    num_tokens = 0
+    for request in documents:
+        num_tokens += len(tokenizer(request["text"],add_special_tokens=False)["input_ids"][0])
+    return num_tokens
+
+@dataclass
+class RequestFuncInput:
+    api_url: str
+    documents: List[str]
+    model: str
+
+
+@dataclass
+class RequestFuncOutput:
+    success: bool = False
+    latency: float = 0.0
+    tokens_len: int = 0
+    error: str = ""
+
+async def async_request_embeddings(
+    request_func_input: RequestFuncInput,
+    pbar: Optional[tqdm] = None,
+) -> RequestFuncOutput:
+    api_url = request_func_input.api_url
+
+    async with aiohttp.ClientSession(timeout=AIOHTTP_TIMEOUT) as session:
+        payload = {
+            "model": request_func_input.model,
+            "input": request_func_input.documents,
+            "encoding_format": "base64",
+        }
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {os.environ.get('OPENAI_API_KEY')}",
+        }
+
+        output = RequestFuncOutput()
+        st = time.perf_counter()
+        try:
+            async with session.post(url=api_url, json=payload,
+                                    headers=headers) as response:
+                if response.status == 200:
+                    async for chunk_bytes in response.content:
+                        if not chunk_bytes:
+                            continue
+                        chunk_bytes = chunk_bytes.decode("utf-8")
+                        data = json.loads(chunk_bytes)
+                        timestamp = time.perf_counter()
+                        output.success = True
+                        output.latency =  timestamp - st
+                else:
+                    output.error = response.reason or ""
+                    output.success = False
+                    print("ERROR",response.reason)
+
+        except Exception:
+            output.success = False
+            exc_info = sys.exc_info()
+            output.error = "".join(traceback.format_exception(*exc_info))
+
+    if pbar:
+        pbar.update(1)
+    return output
+
+async def get_request(
+    documents_all: arrow_dataset.Dataset,
+    request_rate: float,
+) -> AsyncGenerator[List[str], None]:
+    documents = documents_all.iter(batch_size=batch_size)
+    for request in documents:
+        yield request["text"]
+        if request_rate == float("inf"):
+            # If the request rate is infinity, then we don't need to wait.
+            continue
+        # Sample the request interval from the exponential distribution.
+        interval = np.random.exponential(1.0 / request_rate)
+        # The next request will be sent after the interval.
+        await asyncio.sleep(interval)
```

Comment:
Can you add reference what benchmarking code is it based of off? I think it is vLLM's benchmark_serving,py but I cant remember